specs/simple-serialize.md

```diff
@@ -1,424 +1,114 @@
-            for i in range(0, len(lst), items_per_chunk)
-        ]
-    else:
-        # Leave large items alone
-        chunkz = lst
-
-    # Merkleise
-    def next_power_of_2(x):  
-        return 1 if x == 0 else 2**(x - 1).bit_length()
-
-    for i in range(len(chunkz), next_power_of_2(len(chunkz))):
-        chunkz.append(b'\x00' * SSZ_CHUNK_SIZE)
-    while len(chunkz) > 1:     
-        chunkz = [hash(chunkz[i] + chunkz[i+1]) for i in range(0, len(chunkz), 2)]
-
-    # Return hash of root and data length
-    return hash(chunkz[0] + datalen)
+serialized_bytes = ''.join([serialize(element) for element in value])
+assert len(serialized_bytes) < 2**(8 * BYTES_PER_LENGTH_PREFIX)
+serialized_length = len(serialized_bytes).to_bytes(BYTES_PER_LENGTH_PREFIX, 'little')
+return serialized_length + serialized_bytes
 ```
 
-To `hash_tree_root_internal` a list, we simply do:
+## Deserialization
 
-```python
-return merkle_hash([hash_tree_root_internal(item) for item in value])
-```
+Given a type, serialization is an injective function from objects of that type to byte strings. That is, deserialization—the inverse function—is well-defined.
 
-Where the inner `hash_tree_root_internal` is a recursive application of the tree-hashing function (returning less than 32 bytes for short single values).
+## Merkleization
 
-#### Container
+We first define helper functions:
 
-Recursively tree hash the values in the container in the same order as the fields, and Merkle hash the results.
+* `pack`: Given ordered objects of the same basic type, serialize them, pack them into `BYTES_PER_CHUNK`-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.
+* `merkleize`: Given ordered `BYTES_PER_CHUNK`-byte chunks, right-pad them with zero chunks to the next power of two, Merkleize the chunks, and return the root.
+* `mix_in_length`: Given a Merkle root `root` and a length `length` (`uint256` little-endian serialization) return `hash(root + length)`.
 
-```python
-return merkle_hash([hash_tree_root_internal(getattr(x, field)) for field in value.fields])
-```
-
-### Signed roots
+We now define Merkleization `hash_tree_root(value)` of an object `value` recursively:
 
-Let `field_name` be a field name in an SSZ container `container`. We define `truncate(container, field_name)` to be the `container` with the fields from `field_name` onwards truncated away. That is, `truncate(container, field_name) = [getattr(container, field)) for field in value.fields[:i]]` where `i = value.fields.index(field_name)`.
+* `merkleize(pack(value))` if `value` is a basic object or a tuple of basic objects
+* `mix_in_length(merkleize(pack(value)), len(value))` if `value` is a list of basic objects
+* `merkleize([hash_tree_root(element) for element in value])` if `value` is a tuple of composite objects or a container
+* `mix_in_length(merkleize([hash_tree_root(element) for element in value]), len(value))` if `value` is a list of composite objects
```

Comment:
> if `value` is a list of basic objects

I'd personally prefer "if `value` is a list of objects whose serialized values have a length of <32 bytes. The reason is that in the future we may want to serialize arrays of tuples of small variables and have them be packed efficiently; for one example of this see the [CBC proposal](https://github.com/ethereum/eth2.0-specs/issues/701) which includes 13-byte containers.

I would say consider something like this:

* `pack`: Given ordered objects of the same type, serialize them, right-pad each object with zero bytes to the next power of 2 if the size of the object is not a power of 2, pack them into `BYTES_PER_CHUNK`-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.

We now define Merkleization `hash_tree_root(value)` of an object `value` recursively: 

* `merkleize(pack(value))` if `value` is an object or tuple of objects where each object has serialized size <= 32 bytes.
 * `mix_in_length(merkleize(pack(value)), len(value))` if `value` is a list of objects where each object has serialized size <= 32 bytes.
* `merkleize([hash_tree_root(element) for element in value])` if value is a tuple of objects with serialized size > 32 bytes or a container.
* `mix_in_length(merkleize([hash_tree_root(element) for element in value]), len(value))` if `value` is a list of objects with serialized size > 32 bytes.

Reply:
<img width="499" alt="screenshot 2019-02-28 at 11 28 25" src="https://user-images.githubusercontent.com/731743/53560127-fc1f3380-3b4b-11e9-89b2-d89442516a79.png">